tools/windows/DatadogAgentInstaller/CustomActions/PythonDistributionCustomAction.cs

```diff
@@ -225,23 +255,70 @@ private static ActionResult RunPythonScript(ISession session, string script)
             if (string.IsNullOrEmpty(installPythonThirdPartyDeps) || installPythonThirdPartyDeps != "1")
             {
-                sessionWrapper.Log("Skipping installation of third-party Python deps. Set INSTALL_PYTHON_THIRD_PARTY_DEPS=1 to enable this feature.");
+                _session.Log("Skipping installation of third-party Python deps. Set INSTALL_PYTHON_THIRD_PARTY_DEPS=1 to enable this feature.");
                 return ActionResult.Success;
             }
+            return RunPythonScript(_session, "post.py");
 
-            return RunPythonScript(sessionWrapper, "post.py");
+        }
+
+        private ActionResult RunPreRemovePythonScript()
+        {
+            // add the .post_python_installed_packages.txt to the rollback data store
+            // This means the file can be restored on failure and still used to create a diff file for next retry
+            // Helps prevents silent failure of third party integration feature. Where there is no post file so it silenly continues.
+            var pythonPackagesFile = Path.Combine(_session.Property("APPLICATIONDATADIRECTORY"), "protected", ".post_python_installed_packages.txt");
+            // check that file path exists
+            if (!System.IO.File.Exists(pythonPackagesFile))
+            {
+                _session.Log($"File {pythonPackagesFile} does not exist, skipping rollback data store addition.");
+            }
+            else
+            {
+                _rollbackDataStore.Add(new FileStorageRollbackData(pythonPackagesFile));
+            }
+
+            var result = RunPythonScript(_session, "pre.py");
+            _rollbackDataStore.Store();
+            return result;
+        }
+        private void RunRollbackDataRestore()
+        {
+            try
+            {
+                _rollbackDataStore.Restore();
+            }
+            catch (Exception e)
+            {
+                _session.Log($"Error while restoring rollback post install file: {e}");
+            }
+        }
+
+        private ActionResult RunPreRemovePythonScriptRollback()
+        {
+            RunRollbackDataRestore();
+            return ActionResult.Success;
+        }
+
+        public static ActionResult RunPostInstPythonScript(Session session)
+        {
+            return new PythonDistributionCustomAction(new SessionWrapper(session), "pythonDistribution").RunPostInstPythonScript();
         }
 
         public static ActionResult RunPreRemovePythonScript(Session session)
         {
-            return RunPythonScript(new SessionWrapper(session), "pre.py");
+            return new PythonDistributionCustomAction(new SessionWrapper(session), "pythonDistribution").RunPreRemovePythonScript();
+        }
+
+        public static ActionResult RunPreRemovePythonScriptRollback(Session session)
+        {
+            return new PythonDistributionCustomAction(new SessionWrapper(session), "pythonDistribution").RunPreRemovePythonScriptRollback();
```

Comment:
not an issue right now since postinst doesn't use the rollbackdata store, but if it did, then it would overwrite the `pythonDistribution.json` file since you gave them the same name. only the rollback/action pair should use the same store name.